README.md
<h3 align="center">
  <img src="assets/slitherin_icon_web.png" width="300">
</h3>

# Slitherin


AI research environment for the game of Snake written in Python 2.7. Part of the [OpenAI - Request For Research 2.0](https://blog.openai.com/requests-for-research-2/).


Check out corresponding Medium articles:

[Slitherin - Solving the Classic Game of Snake🐍 with AI🤖 (Part 1: Domain Specific Solvers)](https://towardsdatascience.com/slitherin-solving-the-classic-game-of-snake-with-ai-part-1-domain-specific-solvers-d1f5a5ccd635)

[Slitherin - Solving the Classic Game of Snake🐍 with AI🤖 (Part 2: General Purpose Solvers)](https://towardsdatascience.com/slitherin-solving-the-classic-game-of-snake-with-ai-part-2-general-purpose-random-monte-25dc0dd4c4cf)

[Slitherin - Solving the Classic Game of Snake🐍 with AI🤖 (Part 3: Genetic Evolution)](https://towardsdatascience.com/slitherin-solving-the-classic-game-of-snake-with-ai-part-3-genetic-evolution-33186e6be110)


Table of Contents
=================

  * [Usage](#usage)
  * [Rules](#rules)
  * [Modes](#modes)
     * [Domain specific](#domain-specific)
        * [Shortest Path BFS](#shortest-path-bfs)
        * [Shortest Path DFS](#shortest-path-dfs)
        * [Longest path](#longest-path)
        * [Hamilton](#hamilton)
        * [DNN](#dnn)
        * [DNN Monte Carlo](#dnn-monte-carlo)
     * [General purpose](#general-purpose)
        * [Human](#human)
        * [Random](#random)
        * [Monte Carlo](#monte-carlo)
        * [DNN Genetic Evolution](#dnn-genetic-evolution)

## Usage

1. Clone the repo.
2. Go to the project's root folder.
3. Install required packages`pip3 install -r requirements.txt`.
4. Launch slitherin. I recommend starting with the help mode to see all available modes `python slitherin.py --help`.

## Rules
1. Snake has to move either forward, left or right.
2. Snake dies when hits wall or itself.
3. For every eaten fruit, snake's length increases by 1 and a new fruit is generated on a random unoccupied place.

## Modes
All mode previews contain <span style="color:green">current score</span> **Mode Name (min/avg/max)**.
All modes are benchmarked on a 12x12 grid.

### Domain specific
> Algorithms are using domain specific data like snake's position, direction, neighbors etc.


#### Shortest Path BFS
`python3 slitherin.py --shortest_path_bfs`

<img src="assets/gifs/shortest_path_bfs.gif" width="200">

Generates the shortest path from the snake’s head to the fruit using BFS algorithm.

Optimal performance during early stages, but as the snake grows, its body creates an unavoidable obstacle for the leading head. 

---

#### Shortest Path DFS
`python3 slitherin.py --shortest_path_dfs`

<img src="assets/gifs/shortest_path_dfs.gif" width="200">

Generates the shortest path from the snake’s head to the fruit using DFS algorithm.

Performs worse than BFS due to the graph’s cyclicity.

---

#### Longest path
`python3 slitherin.py --longest_path`

<img src="assets/gifs/longest_path.gif" width="200">

Firstly, generates the shortest path (BFS) between the snake’s head and the fruit. Then for each pair of points in the path, tries to extend the distance between them with available actions.

Snake dies when its body is on a generated path.

---

#### Hamilton
`python3 slitherin.py --hamilton`

<img src="assets/gifs/hamilton.gif" width="200">

Generates a longest path between the snake’s head and its tail. 

In the vast majority of the cases, such path covers the whole environment creating [Hamiltonian path](https://en.wikipedia.org/wiki/Hamiltonian_path), thus solving the game of snake with a perfect score.

---

#### DNN

> Each Deep Neural Net mode has a same model structure of:
> 
> * input layer with 5 neurons [action\_vector, left\_neighbor, forward\_neighbor, right\_neighbor, angle\_to\_fruit]
> * hidden layer with 125 neurons (ReLU 6 activation)
> * output layer with 1 neuron (value for a given action\_vector)

`python3 slitherin.py --deep_neural_net`

`python3 slitherin.py --deep_neural_net_trainer`

<img src="assets/gifs/dnn.gif" width="200">

Training phase consists of performing random gameplays followed by the evaluation and backpropagation of performed actions and its results. 

Rewards:

* **0.7** for eating the fruit
* **0.1** for moving towards the fruit
* **-0.2** for moving away of the fruit
* **-1.0** for dying

As expected, DNN solver performs well in the early stages. Snake goes straight to the fruit and doesn't go into cycles. However as it gets longer, it starts to have problems with going around itself. With the current model structure (data about only the nearest surroundings), a snake doesn't indicate any sense of 'the whole environment orientation and position'

---

#### DNN Monte Carlo
`python3 slitherin.py --deep_neural_net_monte_carlo`

<img src="assets/gifs/dnn_monte_carlo.gif" width="200">

For each possible action, there is a DNN-driven gameplay generated. Gameplay with the highest score is chosen for an ultimate move.

Very slow and inefficient performance in the beginning, but favorable in the late stages. DNN-driven simulations allow the snake to choose relatively wise long-term moves.


---

### General purpose
> Algorithms are not using any domain specific data.


#### Human 
`python3 slitherin.py --human`

Used for debug, development and fun:).

---

#### Random 
`python3 slitherin.py --random`

<img src="assets/gifs/random.gif" width="200">


It's always good to start benchmarking against randomness (at least pseudo).

As expected, very low performance.

---

#### Monte Carlo 
`python3 slitherin.py --monte_carlo`

<img src="assets/gifs/monte_carlo.gif" width="200">

For each move, performs a set of 1000 random run simulations. Then groups them by the initial action and finally picks the action that started gameplays with the highest average score.

Slow and weak performance.

---

#### DNN Genetic Evolution
`python3 slitherin.py --deep_neural_net_genetic_evolution`

`python3 slitherin.py --deep_neural_net_genetic_evolution_trainer`

<img src="assets/gifs/dnn_genetic_evolution.gif" width="200">

Initial population starts with random weights. Then in the selection phase, the top 0.1 of the population gets picked to the uniform crossover stage. In the crossover phase, parents are paired using roulette selection (the highest the score, the highest the probability of breeding). Finally, in the mutation phase, 0.01 of the weights of all offsprings are being mutated to the random values. Then we start again with a new population created fully by the newly bred offsprings. Above cycle is being repeated until convergence which happens usually around 25th generation and the average score of 22.

Performance is relatively satisfactory. Snake correctly learned that taking the shortest path to the fruit isn't a good solution in the late stages, but ultimately still gets trapped within its own body.

---

## Original Author
**Greg (Grzegorz) Surma**


CHANGELOG.txt
[April/2/2020]
Slitherin is now Python 3 compatible
-
You can now visualize the model using:
tensorboard --logdir="/tmp/tflearn_logs"
[nothing follows]


game/models/domain_specific/dnn_ai_solver.py

import random
import copy
import numpy as np
from statistics import mean
from game.models.base_game_model import BaseGameModel
from tf_models.dnn_model import DeepNeuralNetModel
from game.helpers.constants import Constants


class BaseDNNGameModel(BaseGameModel):

    model = None

    def __init__(self, long_name, short_name, abbreviation):
        BaseGameModel.__init__(self, long_name, short_name, abbreviation)

    def move(self, environment):
        if self.model is None:
            self.model = DeepNeuralNetModel(Constants.MODEL_DIRECTORY + "dnn/")
        BaseGameModel.move(self, environment)


class DNNSolver(BaseDNNGameModel):

    def __init__(self):
        BaseDNNGameModel.__init__(self, "Deep Neural Net", "deep_neural_net", "dnn")

    def move(self, environment):
        BaseDNNGameModel.move(self, environment)
        predicted_action = self._predict(environment, self.model)
        return predicted_action


class DNNTrainer(BaseDNNGameModel):

    def __init__(self):
        BaseDNNGameModel.__init__(self, "Deep Neural Net", "deep_neural_net_trainer", "dnnt")

    def move(self, environment):
        BaseDNNGameModel.move(self, environment)
        print ("")
        print (self.long_name + ": train and test")
        print ("")
        while True:
            self._train(100)
            self._test(100)

    def _train(self, training_runs):
        training_data = np.array(self._training_observations(training_runs))
        x = np.array([i[0] for i in training_data]).reshape(-1, Constants.MODEL_FEATURE_COUNT, 1)
        y = np.array([i[1] for i in training_data]).reshape(-1, 1)
        self.model.model.fit(x, y)
        self.model.save()

    def _test(self, testing_runs):
        moves_per_run = []
        eat_fruit_moves = 0
        good_moves = 0
        bad_moves = 0
        death_moves = 0

        for i in range(0, testing_runs):
            environment = self.prepare_training_environment(10, 10)
            while True:
                predicted_action = self._predict(environment, self.model)
                pre_distance_from_fruit = environment.distance_from_fruit()
                pre_length = environment.reward()
                if not environment.step(predicted_action) or environment.is_in_fruitless_cycle():
                    death_moves += 1
                    break
                environment.eat_fruit_if_possible()
                post_distance_from_fruit = environment.distance_from_fruit()
                post_length = environment.reward()
                if post_length > pre_length:
                    eat_fruit_moves += 1
                elif post_distance_from_fruit < pre_distance_from_fruit:
                    good_moves += 1
                else:
                    bad_moves += 1
            moves_per_run.append(environment.reward())
        mean_score = mean(moves_per_run)
        print ("Mean score: " + str(mean_score))
        print ("Eat fruit moves: " + str(eat_fruit_moves))
        print ("Good moves: " + str(good_moves))
        print ("Bad moves: " + str(bad_moves))
        print ("Death moves: " + str(death_moves))

    def _training_observations(self, training_runs):
        new_environment = self.prepare_training_environment(10, 10)
        observations = []
        for run_index in range(0, training_runs):
            environment = copy.deepcopy(new_environment)
            observations_for_run = []
            while True:
                random_action = random.choice(environment.possible_actions_for_current_action(environment.snake_action))
                environment_observation = environment.observation(random_action)
                pre_distance_from_fruit = environment.distance_from_fruit()
                pre_length = environment.reward()
                alive = environment.step(random_action)
                if not alive:
                    break
                environment.eat_fruit_if_possible()
                post_distance_from_fruit = environment.distance_from_fruit()
                post_length = environment.reward()

                if post_length > pre_length:
                    observation = [environment_observation, [0.7]]
                elif post_distance_from_fruit < pre_distance_from_fruit:
                    observation = [environment_observation, [0.1]]
                else:
                    observation = [environment_observation, [-0.2]]
                observations_for_run.append(observation)
            observation = [environment_observation, [-1.0]]
            observations_for_run.append(observation)
            observations += observations_for_run
        return observations

